src/rust/catmem/pipe.rs

```diff
@@ -39,36 +59,64 @@ pub struct Pipe {
 
 impl Pipe {
     /// Creates a new pipe.
-    pub fn new(buffer: SharedRingBuffer<u16>) -> Self {
-        Self {
+    pub fn create(name: &str) -> Result<Self, Fail> {
+        Ok(Self {
             state: PipeState::Opened,
-            eof: false,
-            buffer: Rc::new(buffer),
-        }
+            pipe_type: PipeType::Push,
+            buffer: SharedRingBuffer::<u16>::create(name, RING_BUFFER_CAPACITY)?,
+        })
+    }
+
+    pub fn open(name: &str) -> Result<Self, Fail> {
+        Ok(Self {
+            state: PipeState::Opened,
+            pipe_type: PipeType::Pop,
+            buffer: SharedRingBuffer::open(name, RING_BUFFER_CAPACITY)?,
+        })
     }
 
-    /// Sets the end of file flag in the target pipe.
-    pub fn set_eof(&mut self) {
-        self.eof = true;
+    pub fn close(&mut self) -> Result<(), Fail> {
+        self.get_next_state()
     }
 
-    /// Gets the value of he end of file flag in the target pipe.
-    pub fn eof(&self) -> bool {
-        self.eof
+    pub fn try_close(&mut self) -> Result<(), Fail> {
+        match self.buffer.try_enqueue(EOF) {
+            Ok(()) => {
+                self.get_next_state()?;
+                Ok(())
+            },
+            Err(_) => Err(Fail::new(libc::EAGAIN, "Could not push")),
+        }
     }
 
-    /// Gets the state of the pipe.
-    pub fn state(&self) -> PipeState {
-        self.state
+    pub fn try_push(&mut self, byte: &u8) -> bool {
+        let x: u16 = (*byte & 0xff) as u16;
+        self.buffer.try_enqueue(x).is_ok()
     }
 
-    /// Sets the state of the pipe.
-    pub fn set_state(&mut self, state: PipeState) {
-        self.state = state;
+    pub fn try_pop(&mut self) -> (Option<u8>, bool) {
+        if let Some(bytes) = self.buffer.try_dequeue() {
+            let (high, low): (u8, u8) = (((bytes >> 8) & 0xff) as u8, (bytes & 0xff) as u8);
+            if high != 0 {
+                if let Err(_) = self.get_next_state() {
+                    warn!("Already received an EOF on this pipe");
+                }
+            }
+            (Some(low), high != 0)
+        } else {
+            (None, false)
+        }
     }
 
-    /// Gets a reference to the underlying buffer of the target pipe.
-    pub fn buffer(&self) -> Rc<SharedRingBuffer<u16>> {
-        self.buffer.clone()
+    pub fn get_next_state(&mut self) -> Result<(), Fail> {
+        self.state = match self.state {
+            PipeState::Opened => match self.pipe_type {
+                PipeType::Push => PipeState::Closing,
+                PipeType::Pop => PipeState::Closed,
+            },
+            PipeState::Closing => PipeState::Closed,
+            PipeState::Closed => return Err(Fail::new(libc::EINVAL, "Cannot close an already closed pipe")),
```

Comment:
Could you follow the same pattern that we have been using in the code, where we always log an error message when we fail?